### PR TITLE
Lde/reinstate work queue

### LINKED ISSUE #214: Re-introduce work queue
Update Honk so that all commitments are computed via the work queue.

### PR DESCRIPTION
**Pull request by ledwards2225**
# Description

Add a simplified `work_queue` for Honk. Use it to compute commitments throughout StandardHonk prover.

The new work queue for Honk is largely a copy-paste then simplify. The only real changes are
- work queue now owns a commitment key instead of calling pedersen directly. (This is instead of the Prover owning the proving key which now no longer makes sense since it is not responsible for commitments).
- simplified interface for adding things to the work queue, i.e. `add_commitment` instead of `add_to_queue(Lots of ugly details)`
- Remove some objects from `WorkItem` like the vague and no longer needed `index` and `constant` which were needed only for FFT operations

This PR also includes a simple change to the original `work_queue` that allows the MSM size to be set directly and flexibly rather than based on an enum. I made this change when I was still planning on using the original work queue for Honk (to accommodate Fold poly commitments). It's no longer necessary but seems like a nice simplifying change so I've left it in.

# Checklist:

- [x] I have reviewed my diff in github, line by line.
- [x] Every change is related to the PR description.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to the issue(s) that it resolves.
- [x] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [x] There are no circuit changes, OR specifications in `/markdown/specs` have been updated.
- [x] There are no circuit changes, OR a cryptographer has been assigned for review.
- [x] I've updated any terraform that needs updating (e.g. environment variables) for deployment.
- [x] The branch has been rebased against the head of its merge target.
- [x] I'm happy for the PR to be merged at the reviewer's next convenience.
- [x] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [x] If existing code has been modified, such documentation has been added or updated.

### CODE REVIEW

File: cpp/src/barretenberg/honk/proof_system/work_queue.hpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
+#pragma once
```
Comment (codygunton) [inline cpp/src/barretenberg/honk/proof_system/work_queue.hpp:1]: It seems like you're deciding (rightly IMO) that the honk work queue is going to not share code with the plonk work queue. In that case, will you move the work queue from `proof_system` to `plonk`? Otherwise we create tech debt.

File: cpp/src/barretenberg/plonk/proof_system/prover/prover.cpp [hunk 2]
```diff
@@ -79,7 +79,7 @@ template <typename settings> void ProverBase<settings>::compute_wire_commitments
         barretenberg::fr* coefficients = key->polynomial_store.get(wire_tag).get_coefficients();
 
         // This automatically saves the computed point to the transcript
-        fr domain_size_flag = i > 2 ? work_queue::MSMType::MONOMIAL_N : work_queue::MSMType::MONOMIAL_N_PLUS_ONE;
+        fr domain_size_flag = i > 2 ? key->circuit_size : (key->circuit_size + 1);
```
Comment (codygunton) [inline cpp/src/barretenberg/plonk/proof_system/prover/prover.cpp:82]: Right, this is a nice simplification.

File: cpp/src/barretenberg/honk/proof_system/work_queue.hpp [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+#include "barretenberg/plonk/proof_system/proving_key/proving_key.hpp"
+#include <cstddef>
+#include <memory>
+
+namespace proof_system::honk {
+
+// Currently only one type of work queue operation but there will likely be others related to Sumcheck
+enum WorkType { SCALAR_MULTIPLICATION };
+
+template <typename Params> class work_queue {
```
Comment (codygunton) [inline cpp/src/barretenberg/honk/proof_system/work_queue.hpp:16]: Is this the same sort of `Params` expected by some other class? Would be helpful to document that here if so.

File: cpp/src/barretenberg/honk/proof_system/work_queue.hpp [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    };
+
+    struct work_item {
+        WorkType work_type = SCALAR_MULTIPLICATION;
+        std::span<FF> mul_scalars;
+        std::string label;
+    };
+
+  private:
+    std::shared_ptr<proof_system::plonk::proving_key> key;
```
Comment (codygunton) [inline cpp/src/barretenberg/honk/proof_system/work_queue.hpp]: Can we call this `proving_key` since there's another type of key floating around in this class?

File: cpp/src/barretenberg/honk/proof_system/work_queue.hpp [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        return work_item_info{ scalar_mul_count };
+    };
+
+    [[nodiscard]] FF* get_scalar_multiplication_data(size_t work_item_number) const
+    {
+        size_t count = 0;
+        for (const auto& item : work_item_queue) {
+            if (item.work_type == WorkType::SCALAR_MULTIPLICATION) {
+                if (count == work_item_number) {
+                    return const_cast<FF*>(item.mul_scalars.data());
```
Comment (codygunton) [inline cpp/src/barretenberg/honk/proof_system/work_queue.hpp:70]: Why is this needed? I think it's easy to get undefined behavior when using `const_cast`. Can't we just return the span?

File: cpp/src/barretenberg/honk/proof_system/work_queue.hpp [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+                if (count == work_item_number) {
+                    return item.mul_scalars.size();
+                }
+                ++count;
+            }
+        }
+        return 0;
+    };
+
+    void put_scalar_multiplication_data(const Commitment& result, size_t work_item_number)
```
Comment (codygunton) [inline cpp/src/barretenberg/honk/proof_system/work_queue.hpp:92]: plonk passes by value. Are we sure we want a difference here?

File: cpp/src/barretenberg/honk/pcs/kzg/kzg.test.cpp [hunk 7]
```diff
@@ -40,7 +40,8 @@ TYPED_TEST(BilinearAccumulationTest, single)
 
     auto prover_transcript = ProverTranscript<Fr>::init_empty();
 
-    KZG::reduce_prove(this->ck(), opening_pair, witness, prover_transcript);
+    auto quotient_W = KZG::compute_opening_proof_polynomial(opening_pair, witness);
```
Comment (codygunton) [inline cpp/src/barretenberg/honk/pcs/kzg/kzg.test.cpp:43]: Really prefer this because I don't have to think about what the name means.

File: cpp/src/barretenberg/honk/composer/composer_helper/standard_honk_composer_helper.cpp [hunk 8]
```diff
@@ -168,12 +168,6 @@ StandardProver StandardHonkComposerHelper<CircuitConstructor>::create_prover(
     auto manifest = Flavor::create_manifest(circuit_constructor.public_inputs.size(), num_sumcheck_rounds);
     StandardProver output_state(std::move(wire_polynomials), circuit_proving_key);
 
-    // TODO(Cody): This should be more generic
```
Comment (codygunton) [inline cpp/src/barretenberg/honk/composer/composer_helper/standard_honk_composer_helper.cpp:171]: ...becuase now the commitment key lives in the work queue ✔️

**Review by codygunton — APPROVED** [review]
Nice work. Asking for a discussion and for you to resolve some tech debt that this work would create.

File: cpp/src/barretenberg/honk/pcs/kzg/kzg.test.cpp (hunk 7 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/honk/pcs/kzg/kzg.test.cpp:43]: You mean you prefer my change or the original?

File: cpp/src/barretenberg/honk/proof_system/work_queue.hpp (hunk 3 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/honk/proof_system/work_queue.hpp:16]: That's right. Comment/TODO added for a bit more clarity

File: cpp/src/barretenberg/honk/proof_system/work_queue.hpp (hunk 4 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/honk/proof_system/work_queue.hpp]: good call, done

File: cpp/src/barretenberg/honk/proof_system/work_queue.hpp (hunk 5 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/honk/proof_system/work_queue.hpp:70]: Yeah I agree but two things: 1) this function is only used in the WASM/c_bind context where it expects an `fr*` so I didn't want to mess with it too much. 2) It may actually be necessary (at least for now) because the pedersen implementation makes heavy use of pointer arithmetic. (You'll notice that the CommitmentKey is also extracting a raw pointer from a span behind the scenes before passing to perdersen). I'm not sure how hard it would be to update pedersen to use a span somehow but needless to say this is not the time/place for that.

File: cpp/src/barretenberg/honk/proof_system/work_queue.hpp (hunk 6 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/honk/proof_system/work_queue.hpp:92]: It's pretty small so probably not a big deal either way but I did this intentionally because all we're doing is serializing the commitment to the transcript. Its scope ends with the function itself so I think no need to pass by value.

File: cpp/src/barretenberg/honk/proof_system/work_queue.hpp (hunk 1 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/honk/proof_system/work_queue.hpp:1]: good call, done

Comment (codygunton) [reply cpp/src/barretenberg/honk/proof_system/work_queue.hpp:1]: Great, thanks.

File: cpp/src/barretenberg/honk/proof_system/work_queue.hpp (hunk 4 above)
Comment (codygunton) [reply cpp/src/barretenberg/honk/proof_system/work_queue.hpp]: Thanks.

File: cpp/src/barretenberg/honk/proof_system/work_queue.hpp (hunk 5 above)
Comment (codygunton) [reply cpp/src/barretenberg/honk/proof_system/work_queue.hpp:70]: Fair enough, but we should keep this in mind.

File: cpp/src/barretenberg/honk/pcs/kzg/kzg.test.cpp (hunk 7 above)
Comment (codygunton) [reply cpp/src/barretenberg/honk/pcs/kzg/kzg.test.cpp:43]: Yes, it requires much less thinking to  what your code is doing IMO
